### PR TITLE
Blaze Manage Campaigns: Integrate API on the campaign list screen

### PR DESCRIPTION
**Pull request by kean**
Related Issue #20762

Update the new campaign list screen to use `/search` to fetch the campaign. It's a paginated endpoint, so it required a bit of extra work.

### Pagination

- Log in with a user that has more than two pages of Blaze campaigns
- Open the campaign list
- Verify that the first page is loaded
- Scroll to the bottom
- Verify that the spinner is displayed in the table footer
- Verify that if the request succeeds, the new campaigns are appended to the list
- Verify that if the request fails, a footer with an error message and a retry button is displayed
- Verify that after loading all campaigns, no further requests are made

Suggested testing strategy:

- Mock the search endpoint with the following responses: [mock-responses.zip](https://github.com/wordpress-mobile/WordPress-iOS/files/11910349/pages.zip) (page 1, page 2)
- Use proxy breakpoints to simulate the delay
- Use proxy breakpoints to simulate network errors

https://github.com/wordpress-mobile/WordPress-iOS/assets/1567433/7ad0df6e-355b-4f41-b531-9c0bf08f7c91

### Full-to-refresh

- Open the campaign list
- Invoke pull-to-refresh
- Verify that if the request succeeds, the data is refreshed
- Verify that if the request fails, the error message is displayed
- Verify that the spinner is hidden after a short delay

https://github.com/wordpress-mobile/WordPress-iOS/assets/1567433/d92dcd96-6a0b-46dd-aba1-ecaad28cb71e

### Refresh After Campaign Creation

- Open the campaign list
- Tap "Plus" and create a new campaign
- Verify that after returning to the campaign list, the new campaign is displayed (potentially after a small delay)

## More Screenshots

Screenshot 2023-06-29 at 4 22 09 PM

Screenshot 2023-06-29 at 4 22 16 PM

Screenshot 2023-06-29 at 4 27 51 PM

## Regression Notes

1. Potential unintended areas of impact: Blaze Campaign List (new screen)
2. What I did to test those areas of impact (or what existing automated tests I relied on): manual testing
3. What automated tests I added (or what prevented me from doing so): n/a

PR submission checklist:

- [x] I have completed the Regression Notes.
- [ ] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

UI Changes testing checklist:
- [x] Portrait and landscape orientations.
- [x] Light and dark modes.
- [x] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] VoiceOver.
- [x] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn't complete, formatting should still respect the right-to-left layout)
- [x] iPhone and iPad. 
- [ ] Multi-tasking: Split view and Slide over. (iPad)

### CODE REVIEW

File: WordPress/Classes/ViewRelated/Blaze Campaigns/BlazeCampaignsViewController.swift [hunk 1]
```diff
@@ -58,17 +51,111 @@ final class BlazeCampaignsViewController: UIViewController, NoResultsViewHost {
...
+    }
+
+    // MARK: - Actions
+
+    @objc private func buttonRetryTapped() {
+        stream.load()
+    }
+
+    @objc private func setNeedsToRefreshCampaigns() {
+        guard pendingStream == nil else { return }
```
Comment (kean) [inline WordPress/Classes/ViewRelated/Blaze Campaigns/BlazeCampaignsViewController.swift:129]: This line covers the following (rare?) scenario:

- Create a campaign and return to the campaign list; the list starts silently refreshing in the background
- Manually perform PTR

Expected result:
- Only one request is made
- PTR is hidden when is finishes

File: WordPress/Classes/ViewRelated/Site Creation/Shared/UITableView+Header.swift [hunk 2]
```diff
@@ -48,4 +48,31 @@ extension UITableView {
...
+            withHorizontalFittingPriority: .required,
+            verticalFittingPriority: .fittingSizeLevel
+        )
+        let newFrame = CGRect(origin: .zero, size: size)
+        if tableFooterView.frame.height != newFrame.height {
+            tableFooterView.frame = newFrame
+            self.tableFooterView = tableFooterView
+        }
+    }
+
```
Comment (wargcm) [inline WordPress/Classes/ViewRelated/Site Creation/Shared/UITableView+Header.swift:77]: Was this accidentally duplicated from the function above this? It looks like the same code or is there a minor change I'm missing?

File: WordPress/WordPressTest/Bundle+TestExtensions.swift
Comment (wargcm) [inline WordPress/WordPressTest/Bundle+TestExtensions.swift:1]: Nice addition! I'm surprised there wasn't a similar helper already 😆

**Review by wargcm — APPROVED** [review]
✅ Verified everything in the test steps. Only one minor comment, the rest looks good! 👍 

---

I did do a quick accessibility pass on the views. There were a few minor issues that might need addressing but it can be done as a separate PR.

**Large accessibility text**
| Campaign Row | Retry Button |
|----------------|-------------|
| row | retry |

**VoiceOver**
Additionally, I did a quick VoiceOver check and found the following issues:
- Dashboard card doesn't read the impressions or clicks
- It's not clear on the Blaze Campaigns screen that the rows are tappable. Maybe we can add something to the chevron in those cells?

https://github.com/wordpress-mobile/WordPress-iOS/assets/2454408/9e1af5e9-2b6b-4494-b15a-2677a0d409c8

File: WordPress/Classes/ViewRelated/Blaze Campaigns/BlazeCampaignsViewController.swift [hunk 3]
```diff
@@ -182,84 +238,8 @@ private extension BlazeCampaignsViewController {
...
             static let emptyTitle = NSLocalizedString("blaze.campaigns.empty.title", value: "You have no campaigns", comment: "Title displayed when there are no Blaze campaigns to display.")
             static let emptySubtitle = NSLocalizedString("blaze.campaigns.empty.subtitle", value: "You have not created any campaigns yet. Click promote to get started.", comment: "Text displayed when there are no Blaze campaigns to display.")
-            static let errorTitle = NSLocalizedString("Oops", comment: "Title for the view when there's an error loading Blaze campiagns.")
-            static let errorSubtitle = NSLocalizedString("There was an error loading campaigns.", comment: "Text displayed when there is a failure loading Blaze campaigns.")
+            static let errorTitle = NSLocalizedString("blaze.campaigns.errorTitle", value: "Oops", comment: "Title for the view when there's an error loading Blaze campiagns.")
+            static let errorSubtitle = NSLocalizedString("blaze.campaigns.errorMessage", value: "There was an error loading campaigns.", comment: "Text displayed when there is a failure loading Blaze campaigns.")
```
Comment (momo-ozawa) [inline WordPress/Classes/ViewRelated/Blaze Campaigns/BlazeCampaignsViewController.swift:250]: Thanks for fixing this 🙏

File: WordPress/WordPressTest/Dashboard/BlazeCampaignsStreamTests.swift [hunk 4]
```diff
@@ -0,0 +1,269 @@
+import XCTest
+import WordPressKit
+
+@testable import WordPress
+
+final class BlazeCampaignsStreamTests: CoreDataTestCase {
```
Comment (momo-ozawa) [inline WordPress/WordPressTest/Dashboard/BlazeCampaignsStreamTests.swift:6]: 🤩

**Review by momo-ozawa — APPROVED** [review]
Nice work!

**Pull request merged**